aiojobs/_scheduler.py

```diff
@@ -62,12 +62,16 @@ async def spawn(self, coro):
             raise RuntimeError("Scheduling a new job after closing")
         job = Job(coro, self)
         should_start = self._limit is None or self.active_count < self._limit
-        self._jobs.add(job)
         if should_start:
             job._start()
         else:
-            # wait for free slot in queue
-            await self._pending.put(job)
+            try:
+                # wait for free slot in queue
+                await self._pending.put(job)
+            except asyncio.CancelledError:
+                await job.close()
+                raise
+        self._jobs.add(job)
```

Comment:
I wonder if it makes more sense to only add jobs when they are started (i.e. "active")?

The only reason I can see for adding pending jobs, is to clean up all the jobs on `.close()`. But, that code also pumps all the jobs out of the queue anyway.
https://github.com/aio-libs/aiojobs/blob/104b4f614b0ea639dc7ebfd590df2faa98ba261d/aiojobs/_scheduler.py#L82-L92

So, maybe we should only track the "active" jobs in this object?

Reply:
So, in .close(), I'm thinking you could do something along the lines of:
```
def get_all_pending():
    while not self._pending.empty(): 
         yield self._pending.get_nowait()

all_jobs = self._jobs.union(get_all_pending())
```

Reply:
Yes, putting only active jobs in _jobs looks interesting. But is it needed here?
This approach requires fixing not only close() but also __iter__/__contains__/etc.
This seems to be more than just a bug fixing and could be done in another task i think

Reply:
Ah, right. Maybe it should be kept for iter/contains. Then I think the fix looks reasonable.